tests/pipeline/test_blaze.py

```diff
@@ -0,0 +1,449 @@
+        self.assertEqual(len(loader), 1)
+        exprdata = loader[ds]
+        self.assertTrue(exprdata.expr.isidentical(expr.ds))
+        self.assertTrue(exprdata.deltas.isidentical(expr.ds_deltas))
+
+    def test_auto_deltas_fail_warn(self):
+        with warnings.catch_warnings(record=True) as ws:
+            warnings.simplefilter('always')
+            loader = BlazeLoader()
+            expr = bz.Data(self.df, dshape=self.dshape)
+            from_blaze(
+                expr,
+                loader=loader,
+                no_deltas_rule='warn',
+            )
+        self.assertEqual(len(ws), 1)
+        w = ws[0].message
+        self.assertIsInstance(w, NoDeltasWarning)
+        self.assertIn(str(expr), str(w))
+
+    def test_auto_deltas_fail_raise(self):
+        loader = BlazeLoader()
+        expr = bz.Data(self.df, dshape=self.dshape)
+        with self.assertRaises(ValueError) as e:
+            from_blaze(
+                expr,
+                loader=loader,
+                no_deltas_rule='raise',
+            )
+        self.assertIn(str(expr), str(e.exception))
+
+    def test_non_numpy_field(self):
+        expr = bz.Data(
+            [],
+            dshape="""
+            var * {
+                 a: datetime,
+                 asof_date: datetime,
+                 timestamp: datetime,
+            }""",
+        )
+        ds = from_blaze(
+            expr,
+            loader=self.garbage_loader,
+            no_deltas_rule='ignore',
+        )
+        with self.assertRaises(AttributeError):
+            ds.a
+        self.assertIsInstance(object.__getattribute__(ds, 'a'), NonNumpyField)
+
+    def test_non_pipeline_field(self):
+        # NOTE: This test will fail if we ever allow string types in
+        # the Pipeline API. If this happens, change the dtype of the `a` field
+        # of expr to another type we don't allow.
+        expr = bz.Data(
+            [],
+            dshape="""
+            var * {
+                 a: string,
+                 asof_date: datetime,
+                 timestamp: datetime,
+            }""",
+        )
+        ds = from_blaze(
+            expr,
+            loader=self.garbage_loader,
+            no_deltas_rule='ignore',
+        )
+        with self.assertRaises(AttributeError):
+            ds.a
+        self.assertIsInstance(
+            object.__getattribute__(ds, 'a'),
+            NonPipelineField,
+        )
+
+    def test_complex_expr(self):
+        expr = bz.Data(self.df, dshape=self.dshape)
+        # put an Add in the table
+        expr_with_add = bz.transform(expr, value=expr.value + 1)
+
+        # Test that we can have complex expressions with no deltas
+        from_blaze(
+            expr_with_add,
+            deltas=None,
+            loader=self.garbage_loader,
+        )
+
+        with self.assertRaises(TypeError):
+            from_blaze(
+                expr.value + 1,  # put an Add in the column
```

Comment:
This was the simple expr traversal case. When we spoke in person about this you said this was fine. The difference in the transform case is that the transform results in a tabular expression but the addition requires us to walk back up the tree.